Incorrect code causes the parser to crash when trying to rewrite

## Compiler version

Scala 3.6.4/3.7.0-RC1

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
//> using options -rewrite -indent

def program: Int => Int =
    {`1`: Int  =>  5}
```

## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
[error] java.lang.ArrayIndexOutOfBoundsException: Index -3 out of bounds for length 22
[error] dotty.tools.dotc.parsing.Parsers$Parser.testChar(Parsers.scala:740)
[error] dotty.tools.dotc.parsing.Parsers$Parser.testChars(Parsers.scala:745)
[error] dotty.tools.dotc.parsing.Parsers$Parser.bracesToIndented(Parsers.scala:863)
[error] dotty.tools.dotc.parsing.Parsers$Parser.inBracesOrIndented(Parsers.scala:648)
[error] dotty.tools.dotc.parsing.Parsers$Parser.inDefScopeBraces(Parsers.scala:652)
......
```
</details>

It reproduces as
```
scala-cli compile --server=false -S 3.7.0-RC1 tests/rewrites/i22906.scala
```
but not 
```
sbt:scala3> testCompilation i22906

[info] Test dotty.tools.dotc.CompilationTests.rewrites started
[                                        ] completed (0/1, 0 failed, 0s)-- Error: out/rewrites/i22906/rewrites/i22906/i22906.scala:4:15 --------------------------------------------------------
4 |    {`1`: Int  =>  5}
  |               ^
  |               parentheses are required around the parameter of a lambda
  |               This construct can be rewritten automatically under -rewrite -source 3.0-migration.
[                                        ] completed (0/1, 0 failed, 1s)Compilation failed for: 'out/rewrites/i22906/rewrites/i22906/i22906.scala'
[=======================================>] completed (1/1, 1 failed, 1s)
[error] Test dotty.tools.dotc.CompilationTests.rewrites failed: java.lang.AssertionError: Rewrite test failed, took 1.341 sec
[error]     at dotty.tools.vulpix.ParallelTesting$CompilationTest.checkFail(ParallelTesting.scala:1283)
[error]     at dotty.tools.vulpix.ParallelTesting$CompilationTest.checkRewrites(ParallelTesting.scala:1260)
[error]     at dotty.tools.dotc.CompilationTests.rewrites(CompilationTests.scala:93)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
```

I don't know if `vulpix` is folding in other options or if it progressed.

It also fails with whatever I published locally.
```
scala-cli compile --server=false -S 3.7.1-RC1-bin-SNAPSHOT tests/rewrites/i22906.scala
```

I forgot to try nightly. OK yes, fails on 3.nightly.

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -732,16 +732,17 @@ object Parsers {
 
     def testChar(idx: Int, p: Char => Boolean): Boolean = {
       val txt = source.content
-      idx < txt.length && p(txt(idx))
+      idx >= 0 && idx < txt.length && p(txt(idx))
     }
 
     def testChar(idx: Int, c: Char): Boolean = {
       val txt = source.content
-      idx < txt.length && txt(idx) == c
+      idx >= 0 && idx < txt.length && txt(idx) == c
     }
 
     def testChars(from: Int, str: String): Boolean =
-      str.isEmpty ||
+      str.isEmpty
+      ||
       testChar(from, str.head) && testChars(from + 1, str.tail)
 
     def skipBlanks(idx: Int, step: Int = 1): Int =

--- a/compiler/test/dotty/tools/utils.scala
+++ b/compiler/test/dotty/tools/utils.scala
@@ -124,6 +124,7 @@ private val toolArg = raw"(?://|/\*| \*) ?(?i:(${ToolName.values.mkString("|")})
 private val directiveOptionsArg = raw"//> using options (.*)".r.unanchored
 private val directiveJavacOptions = raw"//> using javacOpt (.*)".r.unanchored
 private val directiveTargetOptions = raw"//> using target.platform (jvm|scala-js)".r.unanchored
+private val directiveUnsupported = raw"//> using (scala) (.*)".r.unanchored
 private val directiveUnknown = raw"//> using (.*)".r.unanchored
 
 // Inspect the lines for compiler options of the form
@@ -141,6 +142,7 @@ def toolArgsParse(lines: List[String], filename: Option[String]): List[(String,S
     case directiveOptionsArg(args) => List(("scalac", args))
     case directiveJavacOptions(args) => List(("javac", args))
     case directiveTargetOptions(platform) => List(("target", platform))
+    case directiveUnsupported(name, args) => Nil
     case directiveUnknown(rest) => sys.error(s"Unknown directive: `//> using ${CommandLineParser.tokenize(rest).headOption.getOrElse("''")}`${filename.fold("")(f => s" in file $f")}")
     case _ => Nil
   }

--- a/tests/neg/i22906.check
+++ b/tests/neg/i22906.check
@@ -1,0 +1,6 @@
+Flag -indent set repeatedly
+-- Error: tests/neg/i22906.scala:6:15 ----------------------------------------------------------------------------------
+6 |    {`1`: Int  =>  5} // error
+  |               ^
+  |               parentheses are required around the parameter of a lambda
+  |               This construct can be rewritten automatically under -rewrite -source 3.0-migration.

--- a/tests/neg/i22906.scala
+++ b/tests/neg/i22906.scala
@@ -1,0 +1,6 @@
+//> using options -rewrite -indent
+//> nominally using scala 3.7.0-RC1
+// does not reproduce under "vulpix" test rig, which enforces certain flag sets?
+
+def program: Int => Int =
+    {`1`: Int  =>  5} // error

